### PR TITLE
Implemented ISolutionWorkspaceService

### LINKED ISSUE #5365: Figure out how to list files & react to changes (common root, new files/projects, ignoring build artefacts, etc.)
https://docs.google.com/document/d/1KVHnP-as_VUkiya8VcUbGKQedBOiU7NEydIJnbNaSFk/edit

### PR DESCRIPTION
**Pull request by ugras-ergun-sonarsource**
Fixes #5365

### CODE REVIEW

File: src/Core/IFolderWorkspaceService.cs [hunk 1]
```diff
@@ -27,7 +27,6 @@ public interface IFolderWorkspaceService
         /// <summary>
         /// Returns true/false if the workspace is in Open-As-Folder mode
         /// </summary>
-        /// <remarks>Will always return false in VS2015 as that mode is not supported in 2015.</remarks>
```
Comment (ugras-ergun-sonarsource) [inline src/Core/IFolderWorkspaceService.cs:30]: just realised something we can cleanup

File: src/Integration/LocalServices/SolutionWorkspaceService.cs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+                            {// sometimes random sdk files are included as parts of project items
+                                return null;
+                            }
+                            if (name != null && name.Length > 0 && !Path.IsPathRooted(name))
+                            { // might not be needed at all
+                                name = AbsolutePathFromRelative(name, projectDir);
+                            }
+                            return name;
+                        });
+            // .Where(File.Exists); // too slow, no false positives
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration/LocalServices/SolutionWorkspaceService.cs:117]: why did you leave this?

File: src/Integration/LocalServices/SolutionWorkspaceService.cs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        private IEnumerable<VSConstants.VSITEMID> ChildrenOf(IVsHierarchy hierarchy, VSConstants.VSITEMID rootID)
+        {
+            var result = new List<VSConstants.VSITEMID>(); // this list is not needed
+
+            for (var itemID = FirstChild(hierarchy, rootID);
+                 itemID != VSConstants.VSITEMID.Nil;
+                 itemID = NextSibling(hierarchy, itemID))
+            {
+                result.Add(itemID);
+                result.AddRange(ChildrenOf(hierarchy, itemID)); // TODO get rid of recursion
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration/LocalServices/SolutionWorkspaceService.cs:140]: this should be an easy change or alternatively the comment should be removed

File: src/Integration/LocalServices/SolutionWorkspaceService.cs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+            var hr = project.GetMkDocument((uint)VSConstants.VSITEMID.Root, out path);
+            Debug.Assert(hr == VSConstants.S_OK || hr == VSConstants.E_NOTIMPL, "GetMkDocument failed for project.");
+
+            return path;
+        }
+
+        [ExcludeFromCodeCoverage]
+        private IEnumerable<VSConstants.VSITEMID> ChildrenOf(IVsHierarchy hierarchy, VSConstants.VSITEMID rootID)
+        {
+            var result = new List<VSConstants.VSITEMID>(); // this list is not needed
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration/LocalServices/SolutionWorkspaceService.cs:133]: the comment

File: src/Integration/LocalServices/SolutionWorkspaceService.cs [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+            }
+
+            return VSConstants.VSITEMID.Nil;
+        }
+
+        /// <summary>
+        /// Transforms a relative path to an absolute one based on a specified base folder.
+        /// </summary>
+        [ExcludeFromCodeCoverage]
+        private string AbsolutePathFromRelative(string relativePath, string baseFolderForDerelativization)
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration/LocalServices/SolutionWorkspaceService.cs:174]: Could this be replaced with Path.Combine?

File: src/Integration/LocalServices/SolutionWorkspaceService.cs [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            return GetLoadedProjects(solution)
+                .SelectMany(AllItemsInProject)
+                .Where(x => x != null)
+                .Where(x => x.Contains("\\"))
+                .Where(x => !x.EndsWith("\\"))
+                .Where(x => !x.Contains("\\.nuget\\"))
+                .Where(x => !x.Contains("\\node_modules\\")) // move filtering closer to path extraction to avoid processing unnecessary items
+                .Distinct(); // might not be needed
```
Comment (georgii-borovinskikh-sonarsource) [inline src/Integration/LocalServices/SolutionWorkspaceService.cs:74]: Is it needed?

**Review by georgii-borovinskikh-sonarsource — CHANGES_REQUESTED** [review]

**Pull request merged**